lib/ruby/lib/alces/stack/templater.rb

```diff
@@ -45,99 +46,108 @@ def append(template_file, append_file, template_parameters={})
-          raise "Templater loop count reached. Check parameters for infinite loops" if count > 100
-          tags = template.scan(/<%=[ \w\*\+\-\/\%\(\)\=\!]*%>/)
-          return template if tags.length == 0
-          error_tag = false
-          error_message  = "Could not find value(s) for:"
-          tags.each do |t|
-            t.scan(/_*[[:alpha:]][[:alnum:]_]*/) do |word|
-              if !template_parameters.has_key?(word.to_sym)
-                error_tag = true
-                error_message << "\n  " << t
-                break
-              end
-            end
+        def replace_erb(template, template_parameters={})
+          return ERB.new(template).result(OpenStruct.new(template_parameters).instance_eval {binding})
+        rescue StandardError => e
+          $stderr.puts "Could not parse ERB"
+          $stderr.puts template.to_s
+          $stderr.puts template_parameters.to_s
+          raise e
+        end
+      end
+
+      class Combiner < Handler
+        def initialize(json, hash={})
+          @combined_hash = {
+            hostip: `hostname -i`.chomp,
+            index: 0
+          }
+          @combined_hash.merge!(hash)
+          fixed_nodename = combined_hash[:nodename]
+          fixed_index = combined_hash[:index]
+          @combined_hash.merge!(load_yaml_hash)
+          @combined_hash.merge!(load_json_hash(json))
+          @parsed_hash = parse_combined_hash
+          if parsed_hash[:nodename] != fixed_nodename or parsed_hash[:index] != fixed_index
+            raise HashOverrideError.new(fixed_nodename, fixed_index, @parsed_hash)
+          end
+        end
+        class HashOverrideError < StandardError
+          def initialize(nodename, index, parsed_hash={})
+            msg = "Original nodename: " << nodename.to_s << "\n"
+            msg << "Original index: " << index.to_s << "\n"
+            msg << parsed_hash.to_s << "\n"
+            msg << "YAML, JSON and ERB can not alter the values of nodename and index"
+            super(msg)
           end
-          raise error_message if error_tag
-          return replace_hash(ERB.new(template).result(OpenStruct.new(template_parameters).instance_eval {binding}), count + 1, template_parameters)
-        end
-
-        def show_options(options={})
-          options[:hostip] = "IP address of host node"
-          # Flags
-          none_flag = true
-          print_json = false
-          print_iterator = false
-          puts
-          puts "The following command line parameters are replaced by ERB:"
-          options.each do |key, value|
-            if key.to_s == 'JSON' and value.to_s == 'true'
-              print_json = true
-            elsif key.to_s == 'ITERATOR' and value.to_s == 'true'
-              print_iterator = true
+        end
+
+        attr_reader :combined_hash
+        attr_reader :parsed_hash
+
+        def load_json_hash(json)
+          return {} if !json or json.empty?
```

Comment:
Another way of writing this would be to use the tertiary operator:

```
def load_json_hash(json)
  (json || {}).empty? ? {} : JSON.parse(json, symbolize_names: true)
end
```

(Also, doesn't require the explicit `return`).